### PR TITLE
feat: add NextBackOff callback to RetryOption

### PR DESCRIPTION
**Pull request by xuxife**

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -16,53 +16,101 @@ var DefaultRetryOption = RetryOption{
...
+	backoff.BackOff
+	nextBackOff func(context.Context, RetryEvent, time.Duration) time.Duration
+
+	ctx context.Context
+	e   RetryEvent
+}
+
+func (b *backOffWithEvent) NextBackOff() time.Duration {
+	bkof := b.BackOff.NextBackOff()
+	if b.nextBackOff == nil || bkof == backoff.Stop {
```
Comment (zhongyi-zhang) [inline retry.go:94]: Said, the original backoff stop has higher priority than the nextBackOff. Would user always understand it in that way?

Comment (xuxife) [reply retry.go:94]: the orignal (inner) backoff like "max attempts", or "timeout" fires `Stop`, then the retry is stopped without calling `NextBackOff` callback

Comment (xuxife) [reply retry.go:94]: You're right, we need to clarify this behavior

Comment (zhongyi-zhang) [reply retry.go:94]: Yeah, either logic is okay. As a good framework, we'd better design it less ambiguity.

**Review by zhongyi-zhang — APPROVED** [review]

**Pull request merged**